CONTRIBUTING.md

```diff
@@ -1,57 +1,169 @@
+
+### Commit Message Format
+
+We follow the conventions on [How to Write a Git Commit Message](http://chris.beams.io/posts/git-commit/).
+
+Be sure to include any related GitHub
+issue references in the commit message.  See [GFM
+syntax](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown)
+for referencing issues
+
+### Unit Tests
+
+All contributions must pass the unit tests in the [Travis
+CI](https://travis-ci.org/vmware/vcd-cli/) build, which runs automatically on
+pull requests.
+
+Where feasible new features should also include fast, easily maintainable unit 
+tests using the Python [unittest module]
+(https://docs.python.org/3.6/library/unittest.html).  Check the
+repo for examples of good style. 
+
+## Contribution Example
+
+Here is a tutorial of adding a feature to fix the foo command using
+Github account imahacker.  If you are an experienced git user feel free
+to adapt it to your own work style.
+
+### Fork the Repo
+
+Navigate to the [vcd-cli repo on
+Github](https://github.com/vmware/vcd-cli) and use the 'Fork' button to
+create a forked repository under your Github account.  This gives you a copy 
+of the repo for pull requests back to vcd-cli. 
+
+### Clone and Set Upstream Remote
+
+Make a local clone of the forked repo and add the base vcd-cli repo as
+the upstream remote repository.
 
 ``` shell
+git clone https://github.com/imahacker/vcd-cli
+cd vcd-cli
 git remote add upstream https://github.com/vmware/vcd-cli.git
-git checkout -b my-new-feature master
+```
+
+The last git command prepares your clone to pull changes from the
+upstream repo and push them into the fork, which enables you to keep
+the fork up to date. More on that shortly.
+
+### Make Changes and Commit
+
+Start a new topic branch from the current HEAD position on master and
+commit your feature changes into that branch.  
+
+``` shell
+git checkout -b foo-command-fix-22 master
+# (Make feature changes)
 git commit -a
-git push origin my-new-feature
+git push origin foo-command-fix-22
 ```
 
-### Staying In Sync With Upstream
+It is a git best practice to put work for each new feature in a separate
+topic branch and use git checkout commands to jump between them.  This
+makes it possible to have multiple active pull requests.  We can accept
+pull requests from any branch, so it's up to you how to manage them.
+
+### Stay in Sync with Upstream
 
-When your branch gets out of sync with the vmware/master branch, use the following to update:
+From time to time you'll need to merge changes from the upstream 
+repo so your topic branch stays in sync with other checkins.  To do so
+switch to your topic branch, pull from the upstream repo,
+and push into the fork.  If there are conflicts you'll need to [merge
+them now](https://stackoverflow.com/questions/161813/how-to-resolve-merge-conflicts-in-git). 
 
 ``` shell
-git checkout my-new-feature
+git checkout foo-command-fix-22
 git fetch -a
 git pull --rebase upstream master
```

Comment:
git pull --rebase upstream master --tags

Reply:
It looks as if git pulls tags automatically from remotes, which makes --tags a no-op.  Currently git pull man pages only show the --no-tags, which turns this behavior off.

Reply:
`--tags` is required in `git pull` command to retrieve tags from upstream, otherwise it won't retrieve the latest tag and the Python module will not update its version. I believe the default behavior (without `--tags`) won't bring a tag if the latest commit hasn't been tagged.

Reply:
Thanks to both of you. I'm going to add it for safety though git docs are ambiguous on this point.  Better safe than sorry!